crates/democracy/src/vote_threshold.rs

```diff
@@ -0,0 +1,114 @@
+//! Voting thresholds.
+
+use crate::Tally;
+use codec::{Decode, Encode};
+use scale_info::TypeInfo;
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
+use sp_runtime::traits::{IntegerSquareRoot, Zero};
+use sp_std::ops::{Add, Div, Mul, Rem};
+
+/// A means of determining if a vote is past pass threshold.
+#[derive(Clone, Copy, PartialEq, Eq, Encode, Decode, sp_runtime::RuntimeDebug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub enum VoteThreshold {
+    /// A supermajority of approvals is needed to pass this vote.
+    SuperMajorityApprove,
+    /// A supermajority of rejects is needed to fail this vote.
+    SuperMajorityAgainst,
+    /// A simple majority of approvals is needed to pass this vote.
+    SimpleMajority,
+}
+
+pub trait Approved<Balance> {
+    /// Given a `tally` of votes and a total size of `electorate`, this returns `true` if the
+    /// overall outcome is in favor of approval according to `self`'s threshold method.
+    fn approved(&self, tally: Tally<Balance>, electorate: Balance) -> bool;
+}
+
+/// Return `true` iff `n1 / d1 < n2 / d2`. `d1` and `d2` may not be zero.
+fn compare_rationals<T: Zero + Mul<T, Output = T> + Div<T, Output = T> + Rem<T, Output = T> + Ord + Copy>(
+    mut n1: T,
+    mut d1: T,
+    mut n2: T,
+    mut d2: T,
```

Comment:
Maybe we should check if d1 and d2 are non-zero here, and return an option/result